lib/fable/events.ex

```diff
@@ -105,43 +205,73 @@ defmodule Fable.Events do
+  end
+
+  @doc false
+  @spec emit(Fable.Config.t(), aggregate_source, emit_callback, Ecto.Multi.changes(), Keyword.t()) ::
+          transation_fun
 
-      error ->
-        raise error
+  def emit(config, schema_or_queryable, fun, changes \\ %{}, opts)
+
+  def emit(%{repo: repo} = config, %Ecto.Query{} = queryable, fun, changes, opts)
+      when is_function(fun, 3) do
+    fn ->
+      schema = lock(queryable, repo) || raise "No aggregate found"
+      check_schema_fields(schema)
+      rollback_on_error(repo, handle_events(config, schema, fun.(schema, repo, changes), opts))
     end
   end
 
-  def unconditionally_emit(%{repo: repo} = config, schema_or_queryable, event, opts \\ []) do
-    # Unlike emit/4, we want to emit this event regardless of whether we've been
-    # passed an out of date schema
-    repo.serial(schema_or_queryable, fn schema ->
-      do_emit(config, schema, event, opts)
-    end)
+  def emit(
+        %{repo: repo} = config,
+        %{__meta__: %Ecto.Schema.Metadata{}} = schema,
+        fun,
+        changes,
+        opts
+      )
+      when is_function(fun, 3) do
+    check_schema_fields(schema)
+
+    fn ->
+      schema = lock(schema, repo) || schema
+      rollback_on_error(repo, handle_events(config, schema, fun.(schema, repo, changes), opts))
+    end
   end
 
-  @spec emit(Fable.Config.t(), Ecto.Queryable.t() | Ecto.Schema.t(), Fable.Event.t(), Keyword.t()) ::
-          {:ok, Ecto.Schema.t()} | {:error, term}
-  def emit(config, schema_or_queryable, event, opts \\ [])
-
-  def emit(%{repo: repo} = config, %Ecto.Query{} = queryable, event, opts) do
-    repo.serial(queryable, fn schema ->
-      do_emit(config, schema, event, opts)
+  @doc false
+  @spec emit(
+          Fable.Config.t(),
+          Ecto.Multi.t(),
+          aggregate_source,
+          term(),
+          emit_callback,
+          Keyword.t()
+        ) ::
+          Ecto.Multi.t()
+  def emit(config, multi, schema_or_queryable, name, fun, opts)
+      when is_function(fun, 3) do
+    Ecto.Multi.run(multi, name, fn _repo, changes ->
+      emit(config, schema_or_queryable, fun, changes, opts)
     end)
   end
 
-  def emit(%{repo: repo} = config, schema, event, opts) do
-    # We are very intentionally ignoring the retrieved _schema here. If emit/4 has been called,
-    # then the user is definitely trying to emit an event based on the version of the aggregate
-    # that they know about. It may become out of date, and that should cause an error.
-    repo.serial(schema, fn _schema ->
-      do_emit(config, schema, event, opts)
+  @spec handle_events(Fable.Config.t(), Fable.aggregate(), event_or_events, Keyword.t()) ::
+          {:ok, Fable.aggregate()} | term
+  defp handle_events(config, aggregate, %_{} = event, opts),
+    do: handle_events(config, aggregate, [event], opts)
+
+  defp handle_events(config, aggregate, events, opts) when is_list(events) do
+    reduce_while_ok(aggregate, events, fn aggregate, event ->
```

Comment:
You could eliminate the previous clause by doing `reduce_while_ok(aggregate, List.wrap(events),`